src/teachy_tv.c

```diff
@@ -0,0 +1,1668 @@
+        mov r1, r10\n\
+        ands r0, r1\n\
+        cmp r0, 0\n\
+        beq _0815C18E\n\
+        movs r5, 0\n\
+    _0815C166:\n\
+        lsls r0, r5, 2\n\
+        ldr r7, [sp]\n\
+        adds r0, r7\n\
+        movs r1, 0x7\n\
+        subs r1, r5\n\
+        lsls r1, 2\n\
+        adds r1, r6\n\
+        movs r2, 0x4\n\
+        bl memcpy\n\
+        adds r0, r5, 0x1\n\
+        lsls r0, 24\n\
+        lsrs r5, r0, 24\n\
+        cmp r5, 0x7\n\
+        bls _0815C166\n\
+        ldr r0, [sp]\n\
+        adds r1, r6, 0\n\
+        movs r2, 0x8\n\
+        bl CpuFastSet\n\
+    _0815C18E:\n\
+        movs r5, 0\n\
+        movs r0, 0xF0\n\
+        mov r8, r0\n\
+        movs r1, 0xF\n\
+        mov r12, r1\n\
+    _0815C198:\n\
+        adds r4, r6, r5\n\
+        ldrb r0, [r4]\n\
+        mov r3, r8\n\
+        ands r3, r0\n\
+        cmp r3, 0\n\
+        beq _0815C1B2\n\
+        mov r7, r9\n\
+        adds r2, r7, r5\n\
+        ldrb r1, [r2]\n\
+        mov r0, r12\n\
+        ands r0, r1\n\
+        adds r0, r3\n\
+        strb r0, [r2]\n\
+    _0815C1B2:\n\
+        ldrb r0, [r4]\n\
+        mov r3, r12\n\
+        ands r3, r0\n\
+        cmp r3, 0\n\
+        beq _0815C1CA\n\
+        mov r0, r9\n\
+        adds r2, r0, r5\n\
+        ldrb r1, [r2]\n\
+        mov r0, r8\n\
+        ands r0, r1\n\
+        adds r0, r3\n\
+        strb r0, [r2]\n\
+    _0815C1CA:\n\
+        adds r0, r5, 0x1\n\
+        lsls r0, 24\n\
+        lsrs r5, r0, 24\n\
+        cmp r5, 0x1F\n\
+        bls _0815C198\n\
+        ldr r0, [sp]\n\
+        bl Free\n\
+        adds r0, r6, 0\n\
+        bl Free\n\
+        add sp, 0x4\n\
+        pop {r3-r5}\n\
+        mov r8, r3\n\
+        mov r9, r4\n\
+        mov r10, r5\n\
+        pop {r4-r7}\n\
+        pop {r0}\n\
+        bx r0\n\
+        ");
+}
+#endif
+
+u16 TeachyTvComputePalIndexArrayEntryByMetaTile(u8 *palIndexArrayBuf, u16 metaTile)
+{
+    u32 pal;
+    u16 i;
+    int firstEntry;
+    int temp;
+
+    pal = (u32)(metaTile << 16) >> 28;
+    i = 0;
+    firstEntry = *palIndexArrayBuf;
+    if ( firstEntry != pal )
+    {
+        if ( firstEntry == 0xFF )
+        {
+            *palIndexArrayBuf = pal;
+        }
+        else
+        {
+            while ( 1 )
+            {
+                ++i;
+                if ( i > 0xF )
+                    break;
+                temp = palIndexArrayBuf[i];
+                if ( temp == pal )
+                    break;
+                if ( temp == 0xFF )
+                {
+                    palIndexArrayBuf[i] = pal;
+                    break;
+                }
+            }
+        }
+    }
+    return (0xF - i);
+}
+
+#ifdef NONMATCHING
+void TeachyTvLoadMapPalette(const struct MapData * const mStruct, u8 *palIndexArray)
+{
+    u8 i;
+    struct Tileset *ts;
+
+    for (i = 0; i < 16 && palIndexArray[i] != 0xFF; i++)
+    {
+        ts = *(palIndexArray + i) > 6 ? mStruct->secondaryTileset : mStruct->primaryTileset;
```

Comment:
might not be a ternary

Reply:
I tried if-else before, but it had no effect (we once discussed that in discord)

Reply:
```c

void TeachyTvLoadMapPalette(const struct MapData * mStruct, const u8 * palIndexArray)
{
    u8 i;
    const struct Tileset * ts;
    u16 * dest;

    for (i = 0; i < 16; i++)
    {
        if (palIndexArray[i] == 0xFF)
            break;
        if (palIndexArray[i] > 6)
            dest = (u16 *)mStruct->secondaryTileset->palettes + 0x10 * palIndexArray[i];
        else
            dest = (u16 *)mStruct->primaryTileset->palettes + 0x10 * palIndexArray[i];
        LoadPalette(dest, 0x10 * (15 - i), 0x20);
    }
}
```

Reply:
f

> ```c
> void TeachyTvLoadMapPalette(const struct MapData * mStruct, const u8 * palIndexArray)
> {
>     u8 i;
>     const struct Tileset * ts;
>     u16 * dest;
> 
>     for (i = 0; i < 16; i++)
>     {
>         if (palIndexArray[i] == 0xFF)
>             break;
>         if (palIndexArray[i] > 6)
>             dest = (u16 *)mStruct->secondaryTileset->palettes + 0x10 * palIndexArray[i];
>         else
>             dest = (u16 *)mStruct->primaryTileset->palettes + 0x10 * palIndexArray[i];
>         LoadPalette(dest, 0x10 * (15 - i), 0x20);
>     }
> }
> ```

fixed